app.py

```diff
@@ -129,55 +320,205 @@ def _format_context(self, similar_documents):
+
+    async def _generate_response(self, **kwargs):
+        model = kwargs.get("model")
+
         if model == "openai":
-            return await self._generate_openai_response(query, context, temperature, tokens)
+            return await self._generate_openai_response(**kwargs)
         elif model == "experimental":
-            return await self._generate_experimental_response(
-                query, context, temperature, tokens
-            )
+            return await self._generate_experimental_response(**kwargs)
         elif model == "fun":
-            return await self._generate_openai_response(
-                query, context, temperature, tokens, fun=True, add_context=addl_context
+            return await self._generate_openai_response(fun=True, **kwargs)
+        else:  # Default to Gemini
+            return await self._generate_gemini_response(
+                kwargs["query"], kwargs["context"]
             )
-        else:
-            return await self._generate_gemini_response(query, context)
 
-    async def _generate_openai_response(
-        self, query, context, temperature=0.2, tokens=3000, fun=None, add_context=None
-    ):
+    async def _generate_openai_response(self, **kwargs):
         try:
-            actual_context = (
-                f"Here is the chat context:\n\n{add_context if fun else context}"
-            )
+            query = kwargs.get("query")
+            context = kwargs.get("context")
+            is_drug = kwargs.get("is_drug")
+            output_format = kwargs.get("output_format")
+            fun = kwargs.get("fun", False)
+            actual_context = f'Here is the chat context:\n\n{kwargs.get("addl_content", "") if kwargs.get("fun") else context}'
             actual_query = (
                 query
-                if fun
+                if kwargs.get("fun")
                 else f"Check your context and find out: {query}\n\n{AppConfig.LLM_SUFFIX}"
             )
+
+            messages = [
+                {
+                    "role": "system",
+                    "content": (
+                        AppConfig.LLM_HUMOROUS_PERSONA
+                        if kwargs.get("fun")
+                        else AppConfig.LLM_SYSTEM
+                    ),
+                },
+                {"role": "user", "content": actual_context},
+                {"role": "user", "content": actual_query},
+            ]
+
+            if is_drug:
+                messages.append(
+                    {
+                        "role": "system",
+                        "content": "Generate a detailed drug information document in JSON format, based on the provided context and query. Add as much detail as possible. If the context includes a source, provide it, otherwise come up with a reliable source yourself. Do NOT cite anything from psychonautwiki.org as a source, under any circumstances. Ensure all information is accurate and sourced from reliable data.",
```

Comment:
is this reliable? negative qualifiers like "do not <so-and-so>" are often ignored or interpreted as positive in complex queries i find

Reply:
I'll do some testing